# 변수

데이터를 저장하는 메모리 공간의 이름
- 쉽게 데이터에 접근할 수 있도록 메모리 특정 공간에 이름을 붙인 것
- 내용을 다른 값으로 갱신할 수 있음

변수명 = 저장할 값
- ```a=6```과 같이 선언하면, 메모리 어딘가에 주소값을 할당받아 저장
- 대소문자 구분에 유의

In [1]:
a = 6
b = 4
a = a/b

In [2]:
a

1.5

In [3]:
c = 3 + 2j
d = 4 + 3j
b = c + d

In [4]:
print(b)

(7+5j)


In [5]:
head = "수학"
tail = "프로그래밍"

In [6]:
a = head * 5
b = head + "1"*5
c = (head + "1") * 5
d = (head + "\n") * 5

In [7]:
print(a,b,c,d,sep="\n")

수학수학수학수학수학
수학11111
수학1수학1수학1수학1수학1
수학
수학
수학
수학
수학



In [8]:
f

NameError: name 'f' is not defined

# 프린트(print) 함수


- ```print()``` 함수는 파이썬에서 화면에 출력하는 함수 
- 다양한 종류의 데이터를 출력하거나 문자열을 조합하여 출력하는 등 다양한 용도로 사용

print(value1, value2, ..., sep=' ', end='\n', file=sys.stdout, flush=False)

- value1, value2, ... : 출력하고자 하는 값들을 나열
- sep (optional) : 값들 사이에 들어갈 구분자를 지정. 기본값은 공백.
- end (optional) : 출력의 마지막에 붙는 문자열을 지정. 기본값은 개행 문자(\n).
- file (optional) : 출력할 파일 객체를 지정. 기본값은 표준 출력(sys.stdout).
- flush (optional) : 출력 버퍼를 비울지 여부를 지정. 기본값은 False.

이스케이프 코드(escape code)

- ```\n```: 문자열 안에서 줄을 바꿀 때 사용

- ```\t```: 문자열 사이에 탬 간격을 줄 때 사용

- ```\\```: 문자 \를 그대로 표현할 때 사용

- ```\'```: 작은 따옴표를 그대로 표현할 때 사용

- ```\"```: 큰 따옴표를 그대로 표현할 때 사용

In [9]:
name = "Dongwook Shin"
job = "professor"

print("Hello,", name, "!")
print("You are a", job, ".")

Hello, Dongwook Shin !
You are a professor .


In [10]:
print("one", "two", "three", sep="-")  # 출력: one-two-three

one-two-three


In [11]:
print("Hello")
print("World!")

Hello
World!


In [12]:
print("Hello", end=" ")
print("World!")  # 출력: Hello World!

Hello World!


In [13]:
file = open("output.txt", "w")  # 파일 열기

print("This is written to a file.", file=file)
print("This will be printed immediately.", flush=True)

file.close()  # 파일 닫기

This will be printed immediately.


In [14]:
import time

for i in range(10): 
    print(i, end =' ', flush = True) 
    time.sleep(1) 

0 1 2 3 4 5 6 7 8 9 

In [15]:
for i in range(10):
    print(i, end = ' ')
    time.sleep(1) 

# 터미널에서 실행

0 1 2 3 4 5 6 7 8 9 

# 자료형

## 숫자형(Number)

### 정수형(Integer)

정수형은 소수점 없이 정수 값을 나타내는 자료형으로, 파이썬에서 매우 기본적이고 자주 사용되는 자료형 중 하나

In [16]:
# 덧셈과 뺄셈 연산
a = 10
b = 5
sum_result = a + b
diff_result = a - b

print("Sum:", sum_result)
print("Difference:", diff_result)

Sum: 15
Difference: 5


In [17]:
# 곱셈과 나눗셈 연산
a = 10
b = 2
mul_result = a * b
div_result = a / b

print("Multiplication:", mul_result)
print("Division:", div_result)

Multiplication: 20
Division: 5.0


In [18]:
# 거듭제곱 연산
base = 2
exponent = 5
result = base ** exponent

print("Result:", result)

Result: 32


In [19]:
# 나머지 연산과 몫 연산
dividend = 20
divisor = 3
remainder = dividend % divisor
quotient = dividend // divisor

print("Remainder:", remainder)
print("Quotient:", quotient)

Remainder: 2
Quotient: 6


In [20]:
# 정수형 변환(문자열을 정수로 변환)
num_str = "123"
num_int = int(num_str)

print("Converted Integer:", num_int)

Converted Integer: 123


In [21]:
# 자료형 확인
num = 42
print(type(num)) 

<class 'int'>


### 실수형

실수형은 소수점을 포함한 숫자를 나타내는 자료형으로, 파이썬에서 매우 기본적이고 자주 사용되는 자료형 중 하나

In [22]:
# 덧셈과 뺄셈 연산
a = 3.14
b = 1.5
sum_result = a + b
diff_result = a - b

print("Sum:", sum_result)
print("Difference:", diff_result)

Sum: 4.640000000000001
Difference: 1.6400000000000001


In [23]:
# 곱셈과 나눗셈 연산
a = 2.5
b = 1.2
mul_result = a * b
div_result = a / b

print("Multiplication:", mul_result)
print("Division:", div_result)

Multiplication: 3.0
Division: 2.0833333333333335


In [24]:
# 거듭제곱 연산
base = 2.0
exponent = 3.0
result = base ** exponent

print("Result:", result)

Result: 8.0


In [25]:
# 실수형 변환(정수를 실수로 변환)
num_int = 10
num_float = float(num_int)

print("Converted Float:", num_float)

Converted Float: 10.0


In [26]:
# 자료형 확인
num = 3.14
print(type(num)) 

<class 'float'>


In [27]:
# 반올림 함수 사용
number = 3.141592653589793
rounded = round(number, 2)  # 소수점 둘째자리까지 반올림

print("Rounded:", rounded)

Rounded: 3.14


In [28]:
# 실수형과 정수형 혼합 연산
a = 5
b = 2.5

# 정수형과 실수형의 덧셈
sum_result = a + b

# 정수형과 실수형의 곱셈
mul_result = a * b

print("Sum Result:", sum_result)
print("Multiplication Result:", mul_result)

Sum Result: 7.5
Multiplication Result: 12.5


### 8진수

8진수(octal)을 사용하기 위해서는 0o 또는 0O(숫자 0 + 알파벳 o 또는 O)로 시작

In [29]:
a = 0o157
print(a)

111


0o157 = $1\times 8^2 + 5\times 8^1 + 7\times 8^0 = 111$

### 16진수

16진수(hexadecimal)를 사용하기 위해서는 0x 또는 0X로 시작

- a 또는 A = 10
- b 또는 B = 11
- c 또는 C = 12
- d 또는 D = 13
- e 또는 E = 14
- f 또는 F = 15

In [30]:
a = 0x1AB
print(a)

427


0x1AB = $1\times 16^2 + 10\times 16^1 + 11\times 16^0 = 427$

In [31]:
b = 0Xafc
print(b)

2812


0Xafc = $10\times 16^2 + 15\times 16^1 + 12\times 16^0 = 2812$

### 숫자형 자료의 연산

In [32]:
a = 3
b = 2

In [33]:
# 정수형 변수들의 덧셈 결과
type(a + b)

int

In [34]:
# 정수형 변수들의 뺄셈 결과
type(a - b)

int

In [35]:
# 정수형 변수들의 곱셈 결과
type(a * b)

int

In [36]:
# 정수형 변수들의 나눗셈 결과
type(a / b)

float

In [37]:
# 정수형 변수들의 몫 결과
type(a // b)

int

In [38]:
# 정수형 변수들의 나머지 결과
type(a % b)

int

In [39]:
a = 3
a = a + 1
print(a)

4


In [40]:
a = 3
a += 1
print(a)

4


In [41]:
a = 3
a = a - 1
print(a)

2


In [42]:
a = 3
a -= 1
print(a)

2


In [43]:
a = 3
a = a * 2
print(a)

6


In [44]:
a = 3
a *= 2
print(a)

6


In [45]:
a = 3
a = a / 2
print(a)

1.5


In [46]:
a = 3
a /= 2
print(a)

1.5


In [47]:
a = 2
b = 3.5

In [48]:
# 정수형 변수와 실수형 변수의 덧셈 결과
type(a + b)

float

In [49]:
# 정수형 변수와 실수형 변수의 뺄셈 결과
type(a - b)

float

In [50]:
# 정수형 변수와 실수형 변수의 곱셈 결과
type(a * b)

float

In [51]:
# 정수형 변수와 실수형 변수의 나눗셈 결과
type(a / b)

float

In [52]:
# 정수형 변수와 실수형 변수의 몫 결과
print(a // b)
type(a // b)

0.0


float

In [53]:
# 정수형 변수와 실수형 변수의 몫 결과
print(b // a)
type(b // a)

1.0


float

In [54]:
# 정수형 변수와 실수형 변수의 나머지 결과
print(a % b)
type(a % b)

2.0


float

In [55]:
# 정수형 변수와 실수형 변수의 나머지 결과
print(b % a)
type(b % a)

1.5


float

## 부울형(Bool)

### 부울형

부울형은 True 또는 False 값을 가지는 자료형으로, 논리 연산의 결과나 조건문의 평가 결과 등에 사용

In [56]:
# True와 False 값 직접 할당
is_true = True
is_false = False

print("is_true:", is_true)
print("is_false:", is_false)

is_true: True
is_false: False


### 논리 연산

In [57]:
# 크기 비교

a = 10
b = 5

# 비교 연산 결과를 부울 변수에 할당
is_greater = a > b
is_less = a < b
is_equal = a == b
is_not_equal = a != b
is_greater_equal = a >= b
is_less_equal = a <= b

print("is_greater:", is_greater)
print("is_less:", is_less)
print("is_equal:", is_equal)
print("is_not_equal:", is_not_equal)
print("is_greater_equal:", is_greater_equal)
print("is_less_equal:", is_less_equal)

is_greater: True
is_less: False
is_equal: False
is_not_equal: True
is_greater_equal: True
is_less_equal: False


In [58]:
x = True
y = False

# 논리 AND 연산의 결과
result_and = x and y
print("AND Result:", result_and) 

# 논리 OR 연산의 결과
result_or = x or y
print("OR Result:", result_or) 

# 논리 NOT 연산의 결과
result_not_x = not x
result_not_y = not y
print("NOT x Result:", result_not_x) 
print("NOT y Result:", result_not_y)

AND Result: False
OR Result: True
NOT x Result: False
NOT y Result: True


- all: 모든 항목이 참이면 True 반환
- any: 하나라도 참이면 True 반환

In [59]:
all([True, 1, {3}])

True

In [60]:
all([True, 1, {}])

False

In [61]:
any([True, 1, {}])

True

In [62]:
any([False, 0, {}])

False

## 문자형

### 문자열(String)

문자형은 문자들의 집합으로 이루어진 자료형으로, 따옴표(' ' 또는 " ")로 둘러싸인 문자열을 사용

In [63]:
# 문자열 변수 선언과 출력 (작은 따옴표)
message = 'Hello, world!'
print(message)

Hello, world!


In [64]:
# 문자열 변수 선언과 출력 (큰 따옴표)
message = "Hello, world!"
print(message)

Hello, world!


In [65]:
# 여러 줄의 문자열 (멀티라인 문자열)
multiline_string = '''
Hello,
this is a
multiline string!
'''
print(multiline_string)


Hello,
this is a
multiline string!



In [66]:
# 여러 줄의 문자열 (멀티라인 문자열)
multiline_string = """
Hello,
this is a
multiline string!
"""
print(multiline_string)


Hello,
this is a
multiline string!



문자열에 따옴표를 포함해야 할 경우 다른 조합을 사용

In [67]:
a = 'I'm happy'

SyntaxError: invalid syntax (309457664.py, line 1)

In [68]:
a = "I'm happy"
print(a)

I'm happy


In [69]:
# 백슬래시(\)를 이용
b = 'I\'m happy'
print(b)

I'm happy


In [70]:
c = 'Hello,\tNice to meet you.'
print(c)

Hello,	Nice to meet you.


In [71]:
c = 'Hello,\nNice to meet you.'
print(c)

Hello,
Nice to meet you.


In [72]:
'124' # 숫자형이 아닌 문자열

'124'

In [73]:
"Someone said, \"Math is not easy\". "

'Someone said, "Math is not easy". '

In [74]:
"\usr\library\text"

SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 0-1: truncated \uXXXX escape (964570768.py, line 1)

In [75]:
"\\usr\\library\\text"

'\\usr\\library\\text'

In [76]:
print("\\usr\\library\\text")

\usr\library\text


### 문자열 인덱싱(Indexing)과 슬라이싱(slicing)

#### 인덱싱

문자열은 각각의 문자가 고유의 번호(순서)를 가지고 있으며 이를 활용하여 각각의 문자를 호출할 수 있다.

In [77]:
a = 'Hello'

In [78]:
a[0]

'H'

In [79]:
a[1]

'e'

In [80]:
a[2]

'l'

In [81]:
a[3]

'l'

In [82]:
a[4]

'o'

In [83]:
a[5]

IndexError: string index out of range

In [84]:
a[-0]

'H'

In [85]:
a[-1]

'o'

In [86]:
a[-2]

'l'

In [87]:
a[-3]

'l'

In [88]:
a[-4]

'e'

In [89]:
a[-5]

'H'

In [90]:
a[-6]

IndexError: string index out of range

 H  e  l  l  o  (5개의 문자)
 
 0  1  2  3  4  (0~4)
 
-5(=-0) -4 -3 -2 -1  (-5~-1)

#### 슬라이싱

슬라이싱을 이용하면 여러 개의 문자를 호출할 수 있다.

In [91]:
a[0:3]

'Hel'

In [92]:
a[1:3]

'el'

In [93]:
a[1:]

'ello'

In [94]:
a[:4]

'Hell'

In [95]:
a[:]

'Hello'

In [96]:
a[0:3:2]

'Hl'

슬라이싱으로 문자열 나누기

In [97]:
a = "20230905Tue"
date = a[0:8]
day = a[8:]

In [98]:
date

'20230905'

In [99]:
day

'Tue'

### 문자열 연산

In [100]:
a = 'Hello, '
b = 'World'

문자열을 더해서 연결할 때 더하기(+) 사용

In [101]:
a + b

'Hello, World'

문자열을 반복하고 싶을 때 곱하기(*) 사용

In [102]:
b*3

'WorldWorldWorld'

문자열의 길이

In [103]:
len(a)

7

In [104]:
len(b)

5

문자열의 특정 문자 개수 세기

In [105]:
a.count('l')

2

문자열의 특정 문자가 처음 나온 위치를 반환

In [106]:
a.index('l')

2

문자열 수정

In [107]:
a = 'Helko'

In [108]:
a[3] = 'l'

TypeError: 'str' object does not support item assignment

In [109]:
a = a[0:3] + 'l' + a[4:]

In [110]:
a

'Hello'

In [111]:
a.replace('l','k')

'Hekko'

### 문자열 포맷팅(Formatting)

문자열 포맷 코드

```%d```:	정수(Integer)

```%c```:	문자 1개(character)

```%s```:	문자열(String)

```%f```:	부동소수(floating-point)

```%%```:	문자 % 자체를 사용할 때

In [112]:
"오늘 운동을 %d시간 했다." % 3

'오늘 운동을 3시간 했다.'

In [113]:
"나는 수학프로그래밍 학점이 %c 이다." % 'A'

'나는 수학프로그래밍 학점이 A 이다.'

In [114]:
"오늘 운동을 %s시간 했다." % 3

'오늘 운동을 3시간 했다.'

In [115]:
"나는 수학프로그래밍 학점이 %s 이다." % 'A+'

'나는 수학프로그래밍 학점이 A+ 이다.'

숫자와 포맷 코드 함께 사용하기

In [116]:
# 숫자는 자리수를 의미하며 오른쪽 정렬을 한다.
"%10s" %'수학프로그래밍'

'   수학프로그래밍'

In [117]:
len("%10s" %'수학프로그래밍')

10

In [118]:
# 숫자 앞에 -를 붙이면 왼쪽 정렬을 한다.
"%-10s" %'수학프로그래밍'

'수학프로그래밍   '

In [119]:
len("%-10s" %'수학프로그래밍')

10

In [120]:
"%-10s 열심히!" %'수학프로그래밍'

'수학프로그래밍    열심히!'

In [121]:
len("%-10s 열심히!" %'수학프로그래밍')

15

In [122]:
# 소수점 표현
"%0.4f" % 23.42136234

'23.4214'

In [123]:
"%10.4f" % 23.42136234

'   23.4214'

In [124]:
len("%10.4f" % 23.42136234)

10

In [125]:
"%3.4f" % 23.42136234

'23.4214'

In [126]:
len("%3.4f" % 23.42136234)

7

In [127]:
"%-10.4f" % 23.42136234

'23.4214   '

format 함수를 이용한 포맷팅

In [128]:
'나는 오늘 {}시간 수업을 들었어'.format(6)

'나는 오늘 6시간 수업을 들었어'

In [129]:
'나는 오늘 {}시간 수업을 들었어'.format('여섯')

'나는 오늘 여섯시간 수업을 들었어'

In [130]:
# 2개 이상의 값 넣기
'나는 오늘 {}과목 수업을 {}시간 들었어'.format('수학프로그래밍',2)

'나는 오늘 수학프로그래밍과목 수업을 2시간 들었어'

In [131]:
# 숫자를 이용하여 2개 이상의 값 넣기
'나는 오늘 {0}과목 수업을 {1}시간 들었어'.format('수학프로그래밍',2)

'나는 오늘 수학프로그래밍과목 수업을 2시간 들었어'

In [132]:
# 숫자를 이용하여 2개 이상의 값 넣기
'나는 오늘 {1}과목 수업을 {0}시간 들었어'.format('수학프로그래밍',2)

'나는 오늘 2과목 수업을 수학프로그래밍시간 들었어'

In [133]:
# 왼쪽 정렬
"{0:<10}".format("hi")

'hi        '

In [134]:
# 오른쪽 정렬
"{0:>10}".format("hi")

'        hi'

In [135]:
# 가운데 정렬
"{0:^10}".format("hi")

'    hi    '

In [136]:
# 공백 채우기
"{0:=^10}".format("hi")

'====hi===='

In [137]:
"{0:!<10}".format("hi")

'hi!!!!!!!!'

f 문자열 포맷팅(3.6 버전 이상에서만 사용 가능)

In [138]:
sub = '수학프로그래밍'
hour = 3
f'나는 오늘 {sub}과목 수업을 {hour}시간 들었어'

'나는 오늘 수학프로그래밍과목 수업을 3시간 들었어'

In [139]:
# 수식 계산 가능
f'나는 오늘 {sub}과목 수업을 {hour-1}시간 들었어'

'나는 오늘 수학프로그래밍과목 수업을 2시간 들었어'

## 리스트형(List)

### 리스트형 기본

- 여러 값들을 모아서 저장하는 자료형
- 여러 값들을 쉼표(,)로 구분하고 대괄호([ ])로 감싸서 표현<br>
  리스트명 = [요소1, 요소2, 요소3, ...]

In [140]:
a = [1,2,3,4,5]
print(a)

[1, 2, 3, 4, 5]


In [141]:
b = []
c = list()
d = ['Math', 'is', 'hard']
e = [1, 2, 'math']
f = [1, [1,2], 'math', ['hello,','world!']]
print(b,c,d,e,f,sep='\n')

[]
[]
['Math', 'is', 'hard']
[1, 2, 'math']
[1, [1, 2], 'math', ['hello,', 'world!']]


### 인덱싱

In [142]:
a = [1,2,3]
print(a)

[1, 2, 3]


In [143]:
a[0]

1

In [144]:
a[1]

2

In [145]:
a[2]

3

In [146]:
a[-1]

3

In [147]:
a[0] + a[2]

4

In [148]:
a = [1,2,3,['a','b']]

In [149]:
a[0]

1

In [150]:
a[1]

2

In [151]:
a[2]

3

In [152]:
a[3]

['a', 'b']

In [153]:
a[3][0]

'a'

### 슬라이싱

In [154]:
a = [1,2,3,4,5,6,7]

In [155]:
a[0:2]

[1, 2]

문자열의 슬라이싱과 비교해보기

In [156]:
a = [1,2,3,['a','b','c','d']]

In [157]:
a[2:]

[3, ['a', 'b', 'c', 'd']]

In [158]:
a[3][2:4]

['c', 'd']

### 리스트의 연산

In [159]:
a = [1,2]
b = [3,4,5]
a + b

[1, 2, 3, 4, 5]

In [160]:
a * 3

[1, 2, 1, 2, 1, 2]

In [161]:
len(a)

2

In [162]:
a[1] + 'a'

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [163]:
str(a[1]) + 'a'

'2a'

### 리스트의 수정 및 삭제

In [164]:
a = [1,2,3,4,5]

In [165]:
a[2] = 6
a

[1, 2, 6, 4, 5]

In [166]:
del a[2]
a

[1, 2, 4, 5]

In [167]:
del a[2:]
a

[1, 2]

### 리스트 변수 복사

In [168]:
a = [1,2,3]
b = a
b[2] = 0
print(a,b)

[1, 2, 0] [1, 2, 0]


b = a와 같이 a를 b로 할당하면 a의 값이 b로 전부 입력되는 것이 아니고 b가 a의 주솟값을 참조<br>
(b에 a를 할당하면 값이 할당되는 것이 아니라 같은 메모리 주소를 가지므로 b의 원소를 바뀌면 해당 주소의 값을 바꾸어 a의 원소도 바뀜)

다른 변수에도 영향을 끼치도록 참조막 복사한 것을 얕은 복사라고 함

In [169]:
a = [1,2,3]
b = a[:]
b[2] = 0
print(a,b)

[1, 2, 3] [1, 2, 0]


In [170]:
a = [1,2,3]
b = a.copy()
b[2] = 0
print(a,b)

[1, 2, 3] [1, 2, 0]


### 리스트 함수

- append: 리스트의 맨 뒤에 값을 추가
- sort: 리스트 요소를 순서대로 정렬
- reverse: 리스트 요소의 순서를 역순으로 재배치
- index: 해당 원소가 처음 나오는 인덱스 값을 반환
- insert: a번째 위치에 b를 삽입
- remove: 첫 번째로 나오는 값을 삭제
- pop: 맨 뒤의 값을 리턴하고 해당 값을 삭제
- count: 리스트에 포함된 값의 개수 세기
- extend: 리스트를 더함 (+와 동일)

In [171]:
a = [1,2,3,5,4]

In [172]:
a.append(1)
a

[1, 2, 3, 5, 4, 1]

In [173]:
a.sort()
a

[1, 1, 2, 3, 4, 5]

In [174]:
a.reverse()
a

[5, 4, 3, 2, 1, 1]

In [175]:
a.index(1)

4

In [176]:
a.index(3)

2

In [177]:
a.index(7)

ValueError: 7 is not in list

In [178]:
a.insert(1,5)
a

[5, 5, 4, 3, 2, 1, 1]

In [179]:
a = [1,2,3,1,4,2,5]

In [180]:
a.remove(2)
a

[1, 3, 1, 4, 2, 5]

In [181]:
a.remove(1)
a

[3, 1, 4, 2, 5]

In [182]:
a.pop()

5

In [183]:
a

[3, 1, 4, 2]

In [184]:
a.pop(2)

4

In [185]:
a

[3, 1, 2]

In [186]:
a = [1,2,3,1,4,2,5]
a.count(2)

2

In [187]:
a.extend([1,2])

In [188]:
a

[1, 2, 3, 1, 4, 2, 5, 1, 2]

In [189]:
1 in a

True

In [190]:
0 in a

False

In [191]:
sum(a) # a의 합 구하기

21

In [192]:
a = [3,1,-4,2]
a.sort()
a

[-4, 1, 2, 3]

In [193]:
a = [3,1,-4,2]
sorted(a)

[-4, 1, 2, 3]

In [194]:
a

[3, 1, -4, 2]

In [195]:
sorted(a, key=abs)

[1, 2, 3, -4]

In [196]:
sorted(a, key=abs,reverse=True)

[-4, 3, 2, 1]

## 튜플형(Tuple)

### 튜플형 기본

- 소괄호를 사용하여 정의
- 기호 없이도 사용할 수 있으나 쉼표로 값을 구분
- 읽기만 가능하고 변경할 수 없음
- 인덱싱과 슬라이싱 가능
- 요소가 1개일 때는 요소 뒤에 쉼표(,)를 반드시 붙여야 함

In [197]:
a = ()
b = (1,)
c = (1,2,3)
d = 1,2,3
e = (1, 2, 'math')
f = (1, (1,2), 'math', ['hello,','world!'])
print(a,b,c,d,e,f,sep='\n')

()
(1,)
(1, 2, 3)
(1, 2, 3)
(1, 2, 'math')
(1, (1, 2), 'math', ['hello,', 'world!'])


In [198]:
b = (1,)
type(b)

tuple

In [199]:
b = (1) # 튜플이 아님
type(b)

int

In [200]:
del f[0]

TypeError: 'tuple' object doesn't support item deletion

In [201]:
f[0] = 3

TypeError: 'tuple' object does not support item assignment

### 인덱싱과 슬라이싱

In [202]:
f[0]

1

In [203]:
f[1]

(1, 2)

In [204]:
f[-1]

['hello,', 'world!']

In [205]:
f[1:]

((1, 2), 'math', ['hello,', 'world!'])

In [206]:
f[-1][1]

'world!'

In [207]:
f[-1][1] = 3

In [208]:
f[-1]

['hello,', 3]

In [209]:
f[-1]=3

TypeError: 'tuple' object does not support item assignment

### 튜플의 연산

In [210]:
a = (1,2)
a * 3

(1, 2, 1, 2, 1, 2)

In [211]:
b = (3,4)
a + b

(1, 2, 3, 4)

In [212]:
len(a+b)

4

튜플은 요소값을 변경할 수 없으므로 sort, insert, remove, pop과 같은 내장 함수가 없음

## 집합형 (Set)

### 집합형 기본

- 요소를 콤마(,)로 구분하고 { }로 감싸서 선언
- 문자열, 리스트, 튜플은 set 함수를 이용
- 중복을 허용하지 않음
- 순서가 없음 -> 인덱싱과 슬라이싱 불가

In [213]:
a = {1,2,3}
print(a)
type(a)

{1, 2, 3}


set

In [214]:
a = set([1,2,3])
a

{1, 2, 3}

In [215]:
b = set((1,2,3))
b

{1, 2, 3}

In [216]:
c = set(1,2,3)

TypeError: set expected at most 1 argument, got 3

In [217]:
c = set("Mathematics")
c

{'M', 'a', 'c', 'e', 'h', 'i', 'm', 's', 't'}

In [218]:
c2 = {"Mathematics"}
c2

{'Mathematics'}

In [219]:
c[1]

TypeError: 'set' object is not subscriptable

In [220]:
d = list(c)
d[3:]

['i', 'a', 'c', 't', 'e', 's']

In [221]:
e = tuple(c)
e[3:]

('i', 'a', 'c', 't', 'e', 's')

### 집합형의 연산

In [222]:
a = set([1,2,3,4,5,6])
b = set([2,4,6,8,10])

교집합

In [223]:
a & b

{2, 4, 6}

In [224]:
a.intersection(b)

{2, 4, 6}

합집합

In [225]:
a | b

{1, 2, 3, 4, 5, 6, 8, 10}

In [226]:
a.union(b)

{1, 2, 3, 4, 5, 6, 8, 10}

차집합

In [227]:
a - b

{1, 3, 5}

In [228]:
b - a

{8, 10}

값 추가하기

In [229]:
a.add(7)
a

{1, 2, 3, 4, 5, 6, 7}

In [230]:
a.update([8,9,10])
a

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}

값 제거하기

In [231]:
a.remove(1)
a

{2, 3, 4, 5, 6, 7, 8, 9, 10}

In [232]:
a.remove(set([9,10]))
a

KeyError: {9, 10}

In [233]:
a.remove([9,10])
a

TypeError: unhashable type: 'list'

원소 개수 세기

In [234]:
len(a)

9

## 딕셔너리형(Dictionary)

### 딕셔너리형 기본

- 딕셔너리는 집합과 유사하게 중괄호를 사용하여 선언
- 집합형과 헷갈릴 수 있으므로 dict 함수 이용 권장
- 키(key)와 값(value)을 한 쌍으로 지정
- 키는 고유값으로 인덱스와 같은 기능 - 변경 불가
- 집합처럼 자료를 중복 지정할 수 없음

In [235]:
a = dict()
a

{}

In [236]:
type(a)

dict

In [237]:
a = {'a': 1, 'b': 2, 3: 'c'}
a

{'a': 1, 'b': 2, 3: 'c'}

In [238]:
type(a)

dict

In [239]:
a['a']

1

In [240]:
a['b']

2

In [241]:
a[3]

'c'

In [242]:
b = {['a','b']: 5}

TypeError: unhashable type: 'list'

In [243]:
b = {('a','b'): 5}

In [244]:
b[('a','b')]

5

In [245]:
c = {'abc': 5}

In [246]:
c['abc']

5

딕셔너리 쌍 추가

In [247]:
c[2] = 'def'
c

{'abc': 5, 2: 'def'}

In [248]:
c['name'] = 'Dongwook'
c

{'abc': 5, 2: 'def', 'name': 'Dongwook'}

딕셔너리 요소 삭제

In [249]:
del c[2]
c

{'abc': 5, 'name': 'Dongwook'}

### 딕셔너리 내장 함수

모든 키 출력

In [250]:
c.keys()

dict_keys(['abc', 'name'])

리스트로 저장

In [251]:
c_keys = list(c.keys())
c_keys

['abc', 'name']

모든 값 출력

In [252]:
c.values()

dict_values([5, 'Dongwook'])

In [253]:
c_values = list(c.values())
c_values

[5, 'Dongwook']

키와 값 모든 쌍 출력

In [254]:
c.items()

dict_items([('abc', 5), ('name', 'Dongwook')])

값 가져오기

In [255]:
c.get('abc',10)

5

In [256]:
c.get('pwd',10)

10

In [257]:
c.get('pwd')

In [258]:
print(c.get('pwd'))

None


키와 값 확인

In [259]:
'abc' in c

True

In [260]:
'pwd' in c

False

In [261]:
'abc' in c.keys()

True

In [262]:
'abc' in c.values()

False

In [263]:
'abc' in c.items()

False

In [264]:
('abc', 5) in c.items()

True